library/core/src/num/mod.rs

```diff
@@ -1442,36 +1442,35 @@ pub fn can_not_overflow<T>(radix: u32, is_signed_ty: bool, digits: &[u8]) -> boo
     radix <= 16 && digits.len() <= mem::size_of::<T>() * 2 - is_signed_ty as usize
 }
 
+fn check_positivity(src: &str, is_signed_ty: bool) -> Result<(&[u8], bool), ParseIntError> {
+    // all valid digits are ascii, so we will just iterate over the utf8 bytes
+    // and cast them to chars. .to_digit() will safely return None for anything
+    // other than a valid ascii digit for the given radix, including the first-byte
+    // of multi-byte sequences
+    let src = src.as_bytes();
+
+    match src.get(0) {
+        None => Err(ParseIntError { kind: IntErrorKind::Empty }),
+        Some(b'+' | b'-') if src[1..].is_empty() => {
+            Err(ParseIntError { kind: IntErrorKind::InvalidDigit })
+        }
+        Some(b'+') => Ok((&src[1..], true)),
+        Some(b'-') if is_signed_ty => Ok((&src[1..], false)),
+        Some(_) => Ok((src, true)),
+    }
```

Comment:
Did you consider using slice patterns? Something like:
```suggestion
    match src {
        [] => Err(ParseIntError { kind: IntErrorKind::Empty }),
        [b'+' | b'-'] => Err(ParseIntError { kind: IntErrorKind::InvalidDigit }),
        [b'+', tail @ ..] => Ok((tail, true)),
        [b'-', tail @ ..] if is_signed_ty => Ok((tail, false)),
        _ => Ok((src, true)),
    }
```

Reply:
I did not think about the, thanks for the idea!

Reply:
Actually, after testing with slice patterns, it seems to cause the code to be re-inlined and therefore blows up llvm-lines again... should I undo the outlining and just use slice patterns to see if that reduces llvm-lines?

Reply:
You're welcome to explore that either way, but as I mentioned, we care more about runtime performance than we do llvm-lines. It might be a little faster to compile when outlined, but IMO it's going to hurt if this *doesn't* get inlined during optimization.

Let me know if you want me to queue a perf run -- it will take a few hours each time to get results though.

Reply:
Testing with slice patterns without outlining seems to increase llvm-lines, I'm still going to work on alternatives to this PR but I'm closing it for now because this outlining is too brittle.